File: bench/generated_media_evidence_inspector.py
"""Independent expected renderer/parser for the generated-media inspector.

This module renders only validated identities and scalar bounds.  It never
copies registry payload bytes into the result.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from bench import generated_media_output_registry as registry
from bench import generated_media_producer_transition as transition

Record = dict[str, Any]

SCHEMA = "glacier-generated-media-evidence-inspector-v1"
MAX_ARCHIVE_BYTES = 16 * 1024 * 1024
MAX_EVIDENCE_BYTES = (
    transition.BATCH_BYTES + registry.MAX_ENTRIES * transition.TRANSITION_RECEIPT_BYTES
)

TOP_FIELDS = (
    "schema",
    "verified",
    "lineage",
    "request_epoch",
    "registry_generation",
    "publication_sequence",
    "modality_mask",
    "receipt_count",
    "receipt_table_bytes",
    "registry_archive_bytes",
    "evidence_bytes",
    "generation_plan_sha256",
    "tenant_scope_sha256",
    "metadata_policy_sha256",
    "challenge_sha256",
    "receipt_table_sha256",
    "previous_batch_sha256",
    "registry_manifest_sha256",
    "registry_archive_sha256",
    "batch_sha256",
    "entries",
)

ENTRY_SCALAR_FIELDS = (
    "index",
    "modality",
    "model_kind",
    "completion_kind",
    "producer_generation",
    "producer_ordinal",
    "registry_ordinal",
    "unit_start",
    "unit_count",
    "timeline_start",
    "timeline_end",
    "weights_bytes",
    "model_input_bytes",
    "model_state_before_bytes",
    "model_output_bytes",
    "model_state_after_bytes",
    "materializer_payload_bytes",
    "raw_output_bytes",
    "encoded_payload_bytes",
    "encoding_abi",
    "payload_offset",
    "producer_publication_sequence",
    "completion_sequence",
)

ENTRY_DIGEST_FIELDS = (
    "artifact_manifest_sha256",
    "adapter_descriptor_sha256",
    "support_set_sha256",
    "model_plan_sha256",
    "model_output_sha256",
    "model_transition_or_source_mapping_sha256",
    "model_result_sha256",
    "producer_plan_or_manifest_sha256",
    "media_object_sha256",
    "materializer_implementation_sha256",
    "materializer_execution_sha256",
    "raw_output_sha256",
    "provenance_sha256",
    "publication_result_sha256",
    "producer_final_state_sha256",
    "completion_result_sha256",
    "encoder_implementation_sha256",
    "format_sha256",
    "encoded_payload_sha256",
    "previous_transition_receipt_sha256",
    "producer_projection_sha256",
    "registry_previous_entry_sha256",
    "registry_entry_sha256",
    "transition_receipt_sha256",
)

ENTRY_FIELDS = ENTRY_SCALAR_FIELDS + ENTRY_DIGEST_FIELDS

TOP_DIGEST_FIELDS = (
    "generation_plan_sha256",
    "tenant_scope_sha256",
    "metadata_policy_sha256",
    "challenge_sha256",
    "receipt_table_sha256",
    "previous_batch_sha256",
    "registry_manifest_sha256",
    "registry_archive_sha256",
    "batch_sha256",
)

MODALITY_NAMES = {
    registry.IMAGE_MODALITY: "image",
    registry.AUDIO_MODALITY: "audio",
    registry.VIDEO_MODALITY: "video",
}
MODEL_KIND_NAMES = {
    transition.STATELESS_MODEL: "stateless",
    transition.STATEFUL_MODEL: "stateful",
}
COMPLETION_KIND_NAMES = {
    transition.NO_COMPLETION: "none",
    transition.PLAYBACK_COMPLETION: "playback",
    transition.DISPLAY_COMPLETION: "display",
}


class GeneratedMediaEvidenceInspectorError(ValueError):
    """The rendered inspector document is not canonical."""


def _hex(value: bytes) -> str:
    if type(value) is not bytes or len(value) != 32:
        raise GeneratedMediaEvidenceInspectorError("invalid digest")
    return value.hex()


def expected_document(batch: Record) -> Record:
    """Return field-ordered expected JSON for one validated oracle batch."""

    if type(batch) is not dict:
        raise GeneratedMediaEvidenceInspectorError("invalid batch")
    try:
        previous = batch["previous"]
        validated = transition.decode_batch(
            batch["evidence_bytes"],
            batch["registry"]["archive_bytes"],
            previous,
        )
    except (
        KeyError,
        TypeError,
        transition.GeneratedMediaProducerTransitionError,
    ) as error:
        raise GeneratedMediaEvidenceInspectorError(
            "invalid transition batch"
        ) from error

    header = validated["header"]
    registry_value = validated["registry"]
    entries = [
        _expected_entry(index, receipt, registry_value["entries"][index])
        for index, receipt in enumerate(validated["receipts"])
    ]
    document: Record = {
        "schema": SCHEMA,
        "verified": True,
        "lineage": ("genesis" if header["registry_generation"] == 1 else "successor"),
        "request_epoch": header["request_epoch"],
        "registry_generation": header["registry_generation"],
        "publication_sequence": header["publication_sequence"],
        "modality_mask": header["modality_mask"],
        "receipt_count": header["receipt_count"],
        "receipt_table_bytes": header["receipt_table_bytes"],
        "registry_archive_bytes": len(registry_value["archive_bytes"]),
        "evidence_bytes": len(validated["evidence_bytes"]),
    }
    for field in TOP_DIGEST_FIELDS:
        document[field] = _hex(header[field])
    document["entries"] = entries
    if tuple(document) != TOP_FIELDS:
        raise AssertionError("expected top-level field order drift")
    return document


def _expected_entry(
    index: int,
    receipt: Record,
    registry_entry: Record,
) -> Record:
    value: Record = {
        "index": index,
        "modality": MODALITY_NAMES[receipt["modality"]],
        "model_kind": MODEL_KIND_NAMES[receipt["model_kind"]],
        "completion_kind": COMPLETION_KIND_NAMES[receipt["completion_kind"]],
        "producer_generation": receipt["producer_generation"],
        "producer_ordinal": receipt["producer_ordinal"],
        "registry_ordinal": receipt["registry_ordinal"],
        "unit_start": receipt["unit_start"],
        "unit_count": receipt["unit_count"],
        "timeline_start": receipt["timeline_start"],
        "timeline_end": receipt["timeline_end"],
        "weights_bytes": receipt["weights_bytes"],
        "model_input_bytes": receipt["model_input_bytes"],
        "model_state_before_bytes": receipt["model_state_before_bytes"],
        "model_output_bytes": receipt["model_output_bytes"],
        "model_state_after_bytes": receipt["model_state_after_bytes"],
        "materializer_payload_bytes": receipt["materializer_payload_bytes"],
        "raw_output_bytes": receipt["raw_output_bytes"],
        "encoded_payload_bytes": receipt["encoded_payload_bytes"],
        "encoding_abi": registry_entry["encoding_abi"],
        "payload_offset": registry_entry["payload_offset"],
        "producer_publication_sequence": receipt["producer_publication_sequence"],
        "completion_sequence": receipt["completion_sequence"],
    }
    for field in ENTRY_DIGEST_FIELDS:
        value[field] = _hex(receipt[field])
    if tuple(value) != ENTRY_FIELDS:
        raise AssertionError("expected entry field order drift")
    return value


def render_expected(batch: Record) -> bytes:
    """Encode the expected document as one canonical compact JSON line."""

    return (
        json.dumps(
            expected_document(batch),
            ensure_ascii=True,
            separators=(",", ":"),
        )
        + "\n"
    ).encode("ascii")


@dataclass(frozen=True)
class _Pairs:
    values: tuple[tuple[str, Any], ...]


def parse_rendered(raw: bytes) -> Record:
    """Parse and strictly validate one canonical inspector JSON line."""

    if type(raw) is not bytes or not raw.endswith(b"\n"):
        raise GeneratedMediaEvidenceInspectorError("invalid rendered bytes")
    try:
        decoded = json.loads(
            raw.decode("ascii"),
            object_pairs_hook=lambda pairs: _Pairs(tuple(pairs)),
        )
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise GeneratedMediaEvidenceInspectorError("invalid rendered JSON") from error
    if not isinstance(decoded, _Pairs):
        raise GeneratedMediaEvidenceInspectorError("invalid rendered root")
    document = _convert_object(decoded, TOP_FIELDS)
    entries = document["entries"]
    if type(entries) is not list or not entries:
        raise GeneratedMediaEvidenceInspectorError("invalid entries")
    document["entries"] = [_convert_object(entry, ENTRY_FIELDS) for entry in entries]
    _validate_document(document)
    canonical = (
        json.dumps(
            document,
            ensure_ascii=True,
            separators=(",", ":"),
        )
        + "\n"
    ).encode("ascii")
    if canonical != raw:
        raise GeneratedMediaEvidenceInspectorError("non-canonical rendered JSON")
    return document


def _convert_object(value: Any, fields: tuple[str, ...]) -> Record:
    if not isinstance(value, _Pairs):
        raise GeneratedMediaEvidenceInspectorError("invalid object")
    keys = tuple(key for key, _ in value.values)
    if keys != fields or len(set(keys)) != len(keys):
        raise GeneratedMediaEvidenceInspectorError("invalid field order")
    return {key: item for key, item in value.values}


def _validate_document(document: Record) -> None:
    if (
        document["schema"] != SCHEMA
        or document["verified"] is not True
        or document["lineage"] not in ("genesis", "successor")
        or type(document["receipt_count"]) is not int
        or document["receipt_count"] != len(document["entries"])
        or not 1 <= document["receipt_count"] <= registry.MAX_ENTRIES
        or type(document["registry_archive_bytes"]) is not int
        or not 0 < document["registry_archive_bytes"] <= MAX_ARCHIVE_BYTES
        or type(document["evidence_bytes"]) is not int
        or not 0 < document["evidence_bytes"] <= MAX_EVIDENCE_BYTES
        or document["receipt_table_bytes"]
        != document["receipt_count"] * transition.TRANSITION_RECEIPT_BYTES
        or document["evidence_bytes"]
        != transition.BATCH_BYTES + document["receipt_table_bytes"]
        or not 0 < document["modality_mask"] <= 0x7
        or (document["lineage"] == "genesis") != (document["registry_generation"] == 1)
    ):
        raise GeneratedMediaEvidenceInspectorError("invalid document envelope")
    for field in TOP_FIELDS[3:11]:
        if type(document[field]) is not int or document[field] < 0:
            raise GeneratedMediaEvidenceInspectorError("invalid document scalar")
    for field in TOP_DIGEST_FIELDS:
        _validate_hex(document[field])
    if (document["lineage"] == "genesis") != (
        document["previous_batch_sha256"] == "0" * 64
    ):
        raise GeneratedMediaEvidenceInspectorError("invalid predecessor shape")
    payload_offset = 0
    for index, entry in enumerate(document["entries"]):
        if (
            entry["index"] != index
            or entry["modality"] not in MODALITY_NAMES.values()
            or entry["model_kind"] not in MODEL_KIND_NAMES.values()
            or entry["completion_kind"] not in COMPLETION_KIND_NAMES.values()
            or entry["payload_offset"] != payload_offset
        ):
            raise GeneratedMediaEvidenceInspectorError("invalid entry identity")
        for field in ENTRY_SCALAR_FIELDS:
            if field in (
                "modality",
                "model_kind",
                "completion_kind",
            ):
                continue
            if type(entry[field]) is not int or entry[field] < 0:
                raise GeneratedMediaEvidenceInspectorError("invalid entry scalar")
        for field in ENTRY_DIGEST_FIELDS:
            _validate_hex(entry[field])
        payload_offset += entry["encoded_payload_bytes"]


def _validate_hex(value: Any) -> None:
    if (
        type(value) is not str
        or len(value) != 64
        or any(character not in "0123456789abcdef" for character in value)
    ):
        raise GeneratedMediaEvidenceInspectorError("invalid hex digest")

File: bench/generated_media_evidence_inspector.zig
//! Experimental read-only inspector for generated-media registry/evidence pairs.
//!
//! The command validates every supplied byte before writing semantic output.
//! It renders identities, bounds, and lineage only; encoded payload bytes are
//! never copied to stdout and no model or materializer callback is invoked.

const std = @import("std");
const core = @import("core");

const checkpoint_file = core.continuation_checkpoint_file;
const registry = core.generated_media_output_registry;
const transition = core.generated_media_producer_transition;

pub const max_archive_bytes: usize = 16 * 1024 * 1024;
pub const max_evidence_bytes: usize =
    transition.batch_header_bytes +
    registry.max_entries * transition.transition_receipt_bytes;

const schema = "glacier-generated-media-evidence-inspector-v1";

const Options = struct {
    archive_path: []const u8,
    evidence_path: []const u8,
    previous_archive_path: ?[]const u8,
    previous_evidence_path: ?[]const u8,
};

const RenderEntryV1 = struct {
    receipt: transition.TransitionReceiptV1,
    encoding_abi: u64,
    payload_offset: u64,
};

const RenderDocumentV1 = struct {
    successor: bool,
    archive_bytes: usize,
    evidence_bytes: usize,
    manifest: registry.GeneratedMediaOutputRegistryManifestV1,
    batch: transition.BatchEvidenceV1,
    entries: [registry.max_entries]RenderEntryV1,
    entry_count: usize,
};

pub fn main() void {
    run() catch |err| {
        const stderr = std.fs.File.stderr();
        var buffer: [512]u8 = undefined;
        var writer = std.fs.File.Writer.init(stderr, &buffer);
        writer.interface.print(
            "generated-media-evidence-inspector: {s}\n",
            .{@errorName(err)},
        ) catch {};
        writer.interface.flush() catch {};
        std.process.exit(2);
    };
}

fn run() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const arguments = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, arguments);
    const options = try parseOptions(arguments);

    const archive = try readBoundedFile(
        allocator,
        options.archive_path,
        max_archive_bytes,
    );
    defer allocator.free(archive);
    const evidence = try readBoundedFile(
        allocator,
        options.evidence_path,
        max_evidence_bytes,
    );
    defer allocator.free(evidence);

    const evidence_shape =
        try transition.decodeBatchEvidenceV1(evidence);
    const successor =
        evidence_shape.batch.registry_generation > 1;
    const has_predecessor =
        options.previous_archive_path != null;
    if (successor and !has_predecessor)
        return error.MissingPredecessor;
    if (!successor and has_predecessor)
        return error.UnexpectedPredecessor;

    var current_archive: registry.DecodedArchiveV1 = undefined;
    var validated_evidence: transition.DecodedBatchEvidenceV1 = undefined;

    if (successor) {
        const previous_archive_path =
            options.previous_archive_path orelse
            return error.MissingPredecessor;
        const previous_evidence_path =
            options.previous_evidence_path orelse
            return error.MissingPredecessor;
        const previous_archive = try readBoundedFile(
            allocator,
            previous_archive_path,
            max_archive_bytes,
        );
        defer allocator.free(previous_archive);
        const previous_evidence = try readBoundedFile(
            allocator,
            previous_evidence_path,
            max_evidence_bytes,
        );
        defer allocator.free(previous_evidence);

        const trusted_predecessor =
            try decodeTrustedArchiveViewV1(previous_archive);
        current_archive = try registry.decodeArchiveV1(
            archive,
            trusted_predecessor.previous(),
        );
        validated_evidence =
            try transition.validateSuccessorArchiveAndEvidenceV1(
                .{
                    .registry_archive = current_archive,
                    .evidence = evidence,
                },
                .{
                    .registry_archive = trusted_predecessor,
                    .evidence = previous_evidence,
                },
            );
    } else {
        current_archive = try registry.decodeArchiveV1(
            archive,
            null,
        );
        validated_evidence =
            try transition.validateArchiveAndEvidenceV1(.{
                .registry_archive = current_archive,
                .evidence = evidence,
            });
    }

    const document = try collectDocumentV1(
        successor,
        current_archive,
        validated_evidence,
    );

    const stdout = std.fs.File.stdout();
    var output_buffer: [4096]u8 = undefined;
    var output = std.fs.File.Writer.init(
        stdout,
        &output_buffer,
    );
    try renderDocumentV1(&output.interface, document);
    try output.interface.flush();
}

fn parseOptions(arguments: []const []const u8) !Options {
    var archive_path: ?[]const u8 = null;
    var evidence_path: ?[]const u8 = null;
    var previous_archive_path: ?[]const u8 = null;
    var previous_evidence_path: ?[]const u8 = null;

    var index: usize = 1;
    while (index < arguments.len) {
        const flag = arguments[index];
        index += 1;
        if (index >= arguments.len)
            return error.InvalidArguments;
        const value = arguments[index];
        index += 1;
        if (std.mem.eql(u8, flag, "--archive")) {
            try setOption(&archive_path, value);
        } else if (std.mem.eql(u8, flag, "--evidence")) {
            try setOption(&evidence_path, value);
        } else if (std.mem.eql(
            u8,
            flag,
            "--previous-archive",
        )) {
            try setOption(&previous_archive_path, value);
        } else if (std.mem.eql(
            u8,
            flag,
            "--previous-evidence",
        )) {
            try setOption(&previous_evidence_path, value);
        } else {
            return error.InvalidArguments;
        }
    }
    if ((previous_archive_path == null) !=
        (previous_evidence_path == null))
        return error.IncompletePredecessor;
    return .{
        .archive_path = archive_path orelse
            return error.InvalidArguments,
        .evidence_path = evidence_path orelse
            return error.InvalidArguments,
        .previous_archive_path = previous_archive_path,
        .previous_evidence_path = previous_evidence_path,
    };
}

fn setOption(
    slot: *?[]const u8,
    value: []const u8,
) !void {
    if (slot.* != null or value.len == 0)
        return error.InvalidArguments;
    slot.* = value;
}

fn readBoundedFile(
    allocator: std.mem.Allocator,
    path: []const u8,
    ceiling: usize,
) ![]u8 {
    var file = if (std.fs.path.isAbsolute(path))
        try std.fs.openFileAbsolute(path, .{})
    else
        try std.fs.cwd().openFile(path, .{});
    defer file.close();

    const before = try file.stat();
    if (before.kind != .file) return error.InvalidInputKind;
    if (before.size == 0) return error.EmptyInput;
    const length = std.math.cast(
        usize,
        before.size,
    ) orelse return error.InputTooLarge;
    if (length > ceiling) return error.InputTooLarge;

    const bytes = try allocator.alloc(u8, length);
    errdefer allocator.free(bytes);
    if (try file.readAll(bytes) != bytes.len)
        return error.InputChanged;
    var probe: [1]u8 = undefined;
    if (try file.pread(&probe, before.size) != 0)
        return error.InputChanged;
    const after = try file.stat();
    if (after.kind != .file or after.size != before.size)
        return error.InputChanged;
    return bytes;
}

/// Build a read-only view for the caller-designated trust anchor.
///
/// The public exact successor validator revalidates this backing archive,
/// binds its evidence, and then checks the current pair against it. Its own
/// predecessor is deliberately outside this one-hop inspector contract.
fn decodeTrustedArchiveViewV1(
    encoded: []const u8,
) !registry.DecodedArchiveV1 {
    var set = try checkpoint_file.decodeSetV1(encoded);
    if (set.object_count != registry.archive_object_count)
        return error.InvalidPredecessorArchive;
    const manifest_object = try set.object(
        .extension,
        registry.manifest_object_ordinal,
    );
    const entry_table_object = try set.object(
        .extension,
        registry.entry_table_object_ordinal,
    );
    const payload_pack_object = try set.object(
        .extension,
        registry.payload_pack_object_ordinal,
    );
    if (manifest_object.abi_version != registry.manifest_abi or
        entry_table_object.abi_version !=
            registry.entry_table_abi or
        payload_pack_object.abi_version !=
            registry.payload_pack_abi)
        return error.InvalidPredecessorArchive;
    return .{
        .archive_bytes = encoded,
        .archive_sha256 = set.checkpoint_sha256,
        .manifest = try registry.decodeManifestV1(
            manifest_object.bytes,
        ),
        .entry_table = entry_table_object.bytes,
        .payload_pack = payload_pack_object.bytes,
    };
}

fn collectDocumentV1(
    successor: bool,
    archive: registry.DecodedArchiveV1,
    evidence: transition.DecodedBatchEvidenceV1,
) !RenderDocumentV1 {
    const count = std.math.cast(
        usize,
        evidence.batch.receipt_count,
    ) orelse return error.InvalidEntryCount;
    if (count == 0 or count > registry.max_entries)
        return error.InvalidEntryCount;
    var document: RenderDocumentV1 = .{
        .successor = successor,
        .archive_bytes = archive.archive_bytes.len,
        .evidence_bytes = evidence.encoded.len,
        .manifest = archive.manifest,
        .batch = evidence.batch,
        .entries = undefined,
        .entry_count = count,
    };
    for (0..count) |index| {
        const receipt = try evidence.receipt(index);
        const entry = try archive.entry(index);
        document.entries[index] = .{
            .receipt = receipt,
            .encoding_abi = entry.encoding_abi,
            .payload_offset = entry.payload_offset,
        };
    }
    return document;
}

fn renderDocumentV1(
    writer: *std.Io.Writer,
    document: RenderDocumentV1,
) !void {
    const batch = document.batch;
    try writer.print(
        "{{\"schema\":\"{s}\",\"verified\":true," ++
            "\"lineage\":\"{s}\",\"request_epoch\":{d}," ++
            "\"registry_generation\":{d}," ++
            "\"publication_sequence\":{d}," ++
            "\"modality_mask\":{d},\"receipt_count\":{d}," ++
            "\"receipt_table_bytes\":{d}," ++
            "\"registry_archive_bytes\":{d}," ++
            "\"evidence_bytes\":{d}",
        .{
            schema,
            if (document.successor) "successor" else "genesis",
            batch.request_epoch,
            batch.registry_generation,
            batch.publication_sequence,
            batch.modality_mask,
            batch.receipt_count,
            batch.receipt_table_bytes,
            document.archive_bytes,
            document.evidence_bytes,
        },
    );
    try writeDigestField(
        writer,
        "generation_plan_sha256",
        batch.generation_plan_sha256,
    );
    try writeDigestField(
        writer,
        "tenant_scope_sha256",
        batch.tenant_scope_sha256,
    );
    try writeDigestField(
        writer,
        "metadata_policy_sha256",
        batch.metadata_policy_sha256,
    );
    try writeDigestField(
        writer,
        "challenge_sha256",
        batch.challenge_sha256,
    );
    try writeDigestField(
        writer,
        "receipt_table_sha256",
        batch.receipt_table_sha256,
    );
    try writeDigestField(
        writer,
        "previous_batch_sha256",
        batch.previous_batch_sha256,
    );
    try writeDigestField(
        writer,
        "registry_manifest_sha256",
        batch.registry_manifest_sha256,
    );
    try writeDigestField(
        writer,
        "registry_archive_sha256",
        batch.registry_archive_sha256,
    );
    try writeDigestField(
        writer,
        "batch_sha256",
        batch.batch_sha256,
    );
    try writer.writeAll(",\"entries\":[");
    for (document.entries[0..document.entry_count], 0..) |
        entry,
        index,
    | {
        if (index != 0) try writer.writeByte(',');
        try renderEntryV1(writer, index, entry);
    }
    try writer.writeAll("]}\n");
}

fn renderEntryV1(
    writer: *std.Io.Writer,
    index: usize,
    entry: RenderEntryV1,
) !void {
    const receipt = entry.receipt;
    try writer.print(
        "{{\"index\":{d},\"modality\":\"{s}\"," ++
            "\"model_kind\":\"{s}\"," ++
            "\"completion_kind\":\"{s}\"," ++
            "\"producer_generation\":{d}," ++
            "\"producer_ordinal\":{d}," ++
            "\"registry_ordinal\":{d}," ++
            "\"unit_start\":{d},\"unit_count\":{d}," ++
            "\"timeline_start\":{d},\"timeline_end\":{d}," ++
            "\"weights_bytes\":{d}," ++
            "\"model_input_bytes\":{d}," ++
            "\"model_state_before_bytes\":{d}," ++
            "\"model_output_bytes\":{d}," ++
            "\"model_state_after_bytes\":{d}," ++
            "\"materializer_payload_bytes\":{d}," ++
            "\"raw_output_bytes\":{d}," ++
            "\"encoded_payload_bytes\":{d}," ++
            "\"encoding_abi\":{d},\"payload_offset\":{d}," ++
            "\"producer_publication_sequence\":{d}," ++
            "\"completion_sequence\":{d}",
        .{
            index,
            modalityName(receipt.modality),
            modelKindName(receipt.model_kind),
            completionKindName(receipt.completion_kind),
            receipt.producer_generation,
            receipt.producer_ordinal,
            receipt.registry_ordinal,
            receipt.unit_start,
            receipt.unit_count,
            receipt.timeline_start,
            receipt.timeline_end,
            receipt.weights_bytes,
            receipt.model_input_bytes,
            receipt.model_state_before_bytes,
            receipt.model_output_bytes,
            receipt.model_state_after_bytes,
            receipt.materializer_payload_bytes,
            receipt.raw_output_bytes,
            receipt.encoded_payload_bytes,
            entry.encoding_abi,
            entry.payload_offset,
            receipt.producer_publication_sequence,
            receipt.completion_sequence,
        },
    );
    inline for (.{
        .{
            "artifact_manifest_sha256",
            receipt.artifact_manifest_sha256,
        },
        .{
            "adapter_descriptor_sha256",
            receipt.adapter_descriptor_sha256,
        },
        .{
            "support_set_sha256",
            receipt.support_set_sha256,
        },
        .{
            "model_plan_sha256",
            receipt.model_plan_sha256,
        },
        .{
            "model_output_sha256",
            receipt.model_output_sha256,
        },
        .{
            "model_transition_or_source_mapping_sha256",
            receipt.model_transition_or_source_mapping_sha256,
        },
        .{
            "model_result_sha256",
            receipt.model_result_sha256,
        },
        .{
            "producer_plan_or_manifest_sha256",
            receipt.producer_plan_or_manifest_sha256,
        },
        .{
            "media_object_sha256",
            receipt.media_object_sha256,
        },
        .{
            "materializer_implementation_sha256",
            receipt.materializer_implementation_sha256,
        },
        .{
            "materializer_execution_sha256",
            receipt.materializer_execution_sha256,
        },
        .{
            "raw_output_sha256",
            receipt.raw_output_sha256,
        },
        .{
            "provenance_sha256",
            receipt.provenance_sha256,
        },
        .{
            "publication_result_sha256",
            receipt.publication_result_sha256,
        },
        .{
            "producer_final_state_sha256",
            receipt.producer_final_state_sha256,
        },
        .{
            "completion_result_sha256",
            receipt.completion_result_sha256,
        },
        .{
            "encoder_implementation_sha256",
            receipt.encoder_implementation_sha256,
        },
        .{
            "format_sha256",
            receipt.format_sha256,
        },
        .{
            "encoded_payload_sha256",
            receipt.encoded_payload_sha256,
        },
        .{
            "previous_transition_receipt_sha256",
            receipt.previous_transition_receipt_sha256,
        },
        .{
            "producer_projection_sha256",
            receipt.producer_projection_sha256,
        },
        .{
            "registry_previous_entry_sha256",
            receipt.registry_previous_entry_sha256,
        },
        .{
            "registry_entry_sha256",
            receipt.registry_entry_sha256,
        },
        .{
            "transition_receipt_sha256",
            receipt.receipt_sha256,
        },
    }) |field| {
        try writeDigestField(writer, field[0], field[1]);
    }
    try writer.writeByte('}');
}

fn writeDigestField(
    writer: *std.Io.Writer,
    name: []const u8,
    digest: transition.Digest,
) !void {
    const encoded = std.fmt.bytesToHex(digest, .lower);
    try writer.print(",\"{s}\":\"", .{name});
    try writer.writeAll(&encoded);
    try writer.writeByte('"');
}

fn modalityName(value: registry.ModalityV1) []const u8 {
    return switch (value) {
        .image => "image",
        .audio => "audio",
        .video => "video",
    };
}

fn modelKindName(value: transition.ModelKindV1) []const u8 {
    return switch (value) {
        .stateless => "stateless",
        .stateful => "stateful",
    };
}

fn completionKindName(
    value: transition.CompletionKindV1,
) []const u8 {
    return switch (value) {
        .none => "none",
        .playback => "playback",
        .display => "display",
    };
}
